src/compiler/core.ts

```diff
@@ -357,6 +371,20 @@ namespace ts {
         return result;
     }
 
+    export function assign<T1 extends Map<{}>, T2, T3>(t: T1, arg1: T2, arg2: T3): T1 & T2 & T3;

```

Comment:
`T1 & T2` is not really the right type -- and you'll have to change this later once we add "spread types" like `{...T1, ...T2}`. I'd actually just return `any` for now.

The biggest difference is that when property names conflict, `T1 & T2` gives you another intersection type for the property, but spread types give you the correct last-one-wins type for the property.